opencodeblocks/blocks/codeblock.py

```diff
@@ -79,21 +81,102 @@ def init_run_button(self):
-        worker = Worker(self.source_editor.kernel, self.source)
-        worker.signals.stdout.connect(self.handle_stdout)
-        worker.signals.image.connect(self.handle_image)
-        self.source_editor.threadpool.start(worker)
+
+        # Set button text to ...
+        self.run_button.setText("...")
+        self.run_all_button.setText("...")
+
+        # Run code by adding to code to queue
+        code = self.source_editor.text()
+        self.source = code
+        kernel = self.source_editor.kernel
+        kernel.execution_queue.append((self, code))
+        if kernel.busy == False:
+            kernel.run_queue()
+
+    def reset_buttons(self):
+        """Reset the text of the run buttons"""
+        self.run_button.setText(">")
+        self.run_all_button.setText(">>")
+
+    def check_input(self):
+        """ Checks wether a block has connected input blocks """
+        for input_socket in self.sockets_in:
+            for _ in input_socket.edges:
+                return True
+        return False
+
+    def check_output(self):
+        """ Checks wether a block has connected output blocks """
+        for output_socket in self.sockets_out:
+            for _ in output_socket.edges:
+                return True
+        return False
+
+    def run_left(self, already_ran):
+        """
+        Run all of the block's dependencies and then run the block
+
+        Args:
+            already_ran (list): List of blocks that have already been ran
+            This avoids running a cell multiple times when run_right calls
+            run_left multiple times
+
+        Returns:
+            list: List of blocks that have been ran + already_ran
+        """
+        # If no dependencies
+        if not self.check_input():
+            return self.run_code()
+
+        # Create the graph from the scene
+        graph = self.scene().create_graph()
+        # BFS through the input graph
+        edges = bfs_edges(graph, self, reverse=True)
+        # Run the blocks found
+        blocks_to_run = [self] + [v for _, v in edges]
+        for block in blocks_to_run[::-1]:
+            if block not in already_ran:
+                block.run_code()
+                already_ran.append(block)
+        return already_ran
+
+    def run_left_without_args(self):
+        """ Same as run_left without args, for the left button to call """
+        self.run_left([])
+
+    def run_right(self):
+        """ Run all of the output blocks and all their dependencies """
+        # If no output, run left
+        if not self.check_output():
+            return self.run_left([])
+
+        # Same as run_left but instead of running the blocks, we'll use run_left
+        graph = self.scene().create_graph()
+        edges = bfs_edges(graph, self)
+        blocks_to_run = [self] + [v for _, v in edges]
+        already_ran = []
+        for block in blocks_to_run[::-1]:
+            already_ran = block.run_left(already_ran)
```

Comment:
If it's the same, avoid duplication and use a private method (private starts with '_')

Reply:
The two functions have two differences in 6 lines, should we keep them both or create a method?

Reply:
Let's keep them for now we will see after refactors